<a href="https://colab.research.google.com/github/njsteere/Coursera_Capstone/blob/main/Capstone_Week_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Applied Data Science Week 3

##Part 1

Note - if you are here to review/grade this notebook, all project requirements are in one notebook.  I will clearly note where each "to-do" ends, to help with grading/comprehension.


#1) Web Scraping and Data Exploration

To explore the Toronto neighborhood data, we first need to scrape the data.  This course's lab used New York data.  Data for Toronto neighborhoods is an added challenge for us as it is not readily available on the internet.  However, there is a Wikipedia page that includes the needed data.  Here, we will scrape the page for data, then wrangle and clean it, before reading into a Pandas dataframe.

We will be using a tool called BeautifulSoup to scrape the page.  This is a bit of overkill for this lab but a nice chance to practice BeautifulSoup if you are new to scraping like I am.

In [ ]:
#install initial dependencies
 
!pip install bs4
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import requests
from bs4 import BeautifulSoup
import numpy as np
import json
from geopy.geocoders import Nominatim
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans

 

print('Libraries installed')

Libraries installed


In [ ]:
#defining the target url
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
html_data=requests.get(url).text

In [ ]:
#reading target url using BeautifulSoup
soup = BeautifulSoup(html_data, 'html5')
table_contents=[]
table=soup.find('table')
len(table)

2

In [ ]:
#Here's our table.  As you will see it needs cleaning up.
print(table)

<table cellpadding="2" cellspacing="0" rules="all" style="width:100%; border-collapse:collapse; border:1px solid #ccc;">

<tbody><tr>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M1A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top; color:#ccc;">
<p><b>M2A</b><br/><span style="font-size:85%;"><i>Not assigned</i></span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M3A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M4A</b><br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>)</span>
</p>
</td>
<td style="width:11%; vertical-align:top;">
<p><b>M5A</b><br/><span style="font-size:85%;"><a h

In [ ]:
#let's loop through the table and clean up.  
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

print(table_contents)

[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

In [ ]:
#Much better.  Now, we read the table into a Pandas dataframe
df=pd.DataFrame(table_contents)
#Shortening certain borough names for readability
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
#Display the first 5 rows
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [ ]:
#Use df.shape to look at the Dataframe's rows and columns.
df.shape

(103, 3)

In [ ]:
#Use df.describe for a quick overview of our content.  This is optional.
df.describe()

PostalCode     Borough    Neighborhood
count         103         103             103
unique        103          15             103
top           M7R  North York  Enclave of M5E
freq            1          24               1

Our Dataframe has 103 rows and 3 columns.  We have 15 unique boroughs and 103 Postal Codes.

##End of part one.  Submission URL \#1 goes here.

#2)Adding latitude and longitude coordinates

The next step is to add the latitude and longitude coordinates of each neighborhood to the dataframe, so that we can utilize the Foursquare location data.  The Geocoder python package is a free alternative to Google's API, but is unreliable.  We could loop through each postal code until we get a valid set of coordinates, but we have been provided a CSV of coordinates by the instructors which we will now bring in.

In [ ]:
#reading the zip codes into a new df
geo_df = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv', index_col='Postal Code')
geo_df

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476
M1J          43.744734 -79.239476
M1K          43.727929 -79.262029
M1L          43.711112 -79.284577
M1M          43.716316 -79.239476
M1N          43.692657 -79.264848
M1P          43.757410 -79.273304
M1R          43.750072 -79.295849
M1S          43.794200 -79.262029
M1T          43.781638 -79.304302
M1V          43.815252 -79.284577
M1W          43.799525 -79.318389
M1X          43.836125 -79.205636
M2H          43.803762 -79.363452
M2J          43.778517 -79.346556
M2K          43.786947 -79.385975
M2L          43.757490 -79.374714
M2M          43.789053 -79.408493
M2N          43.770120 -79.408493
M2P          43.752758 -79.400049
M2R          43.782736 -79.442259
M3A          43.753259 -79.329656
M3B          43.745906 -79.352188
M3C          43.725900 -79.340923
M3H          43.754328 -79.442259
M3J          43.767980 -79.487262
M3K          43.737473 -79.464763
M3L          43.739015 -79.506944
M3M          43.728496 -79.495697
M3N          43.761631 -79.520999
M4A          43.725882 -79.315572
M4B          43.706397 -79.309937
M4C          43.695344 -79.318389
M4E          43.676357 -79.293031
M4G          43.709060 -79.363452
M4H          43.705369 -79.349372
M4J          43.685347 -79.338106
M4K          43.679557 -79.352188
M4L          43.668999 -79.315572
M4M          43.659526 -79.340923
M4N          43.728020 -79.388790
M4P          43.712751 -79.390197
M4R          43.715383 -79.405678
M4S          43.704324 -79.388790
M4T          43.689574 -79.383160
M4V          43.686412 -79.400049
M4W          43.679563 -79.377529
M4X          43.667967 -79.367675
M4Y          43.665860 -79.383160
M5A          43.654260 -79.360636
M5B          43.657162 -79.378937
M5C          43.651494 -79.375418
M5E          43.644771 -79.373306
M5G          43.657952 -79.387383
M5H          43.650571 -79.384568
M5J          43.640816 -79.381752
M5K          43.647177 -79.381576
M5L          43.648198 -79.379817
M5M          43.733283 -79.419750
M5N          43.711695 -79.416936
M5P          43.696948 -79.411307
M5R          43.672710 -79.405678
M5S          43.662696 -79.400049
M5T          43.653206 -79.400049
M5V          43.628947 -79.394420
M5W          43.646435 -79.374846
M5X          43.648429 -79.382280
M6A          43.718518 -79.464763
M6B          43.709577 -79.445073
M6C          43.693781 -79.428191
M6E          43.689026 -79.453512
M6G          43.669542 -79.422564
M6H          43.669005 -79.442259
M6J          43.647927 -79.419750
M6K          43.636847 -79.428191
M6L          43.713756 -79.490074
M6M          43.691116 -79.476013
M6N          43.673185 -79.487262
M6P          43.661608 -79.464763
M6R          43.648960 -79.456325
M6S          43.651571 -79.484450
M7A          43.662301 -79.389494
M7R          43.636966 -79.615819
M7Y          43.662744 -79.321558
M8V          43.605647 -79.501321
M8W          43.602414 -79.543484
M8X          43.653654 -79.506944
M8Y          43.636258 -79.498509
M8Z          43.628841 -79.520999
M9A          43.667856 -79.532242
M9B          43.650943 -79.554724
M9C          43.643515 -79.577201
M9L          43.756303 -79.565963
M9M          43.724766 -79.532242
M9N          43.706876 -79.518188
M9P          43.696319 -79.532242
M9R          43.688905 -79.554724
M9V          43.739416 -79.588437
M9W          43.706748 -79.594054

In [ ]:
#csv has the postal codes in order, so we will go ahead and sort our existing dataframe by postal code as well
df = df.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(', '.join).reset_index()
#df.head()

In [ ]:
#joining the two dataframes together based on PostalCode
df = df.join(geo_df, on='PostalCode')
df.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [ ]:
df.head(
)

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#End of Part 2.  Second URL submission for grading would go here.

#3) - Loading and exploring

Now that we have our neighborhoods with geographical coordinates, we can create a map of the city with neighborhoods superimposed on top.  



In [ ]:
neighborhoods = df[['Borough', 'Neighborhood', 'Latitude', 'Longitude']]
neighborhoods.head()

Borough                            Neighborhood   Latitude  Longitude
0  Scarborough                          Malvern, Rouge  43.806686 -79.194353
1  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711
3  Scarborough                                  Woburn  43.770992 -79.216917
4  Scarborough                               Cedarbrae  43.773136 -79.239476

In [ ]:

address = 'Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

#import folium
import folium # map rendering library

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


#From here, we can explore the data in lots of ways.

For demonstration purposes, let's suppose we just wanted to look at neighborhoods in the Borough of Scarborough.  So, let's slice our dataframe and create a new dataframe of just the Scarborough data.

In [ ]:
scarborough_data = neighborhoods[neighborhoods['Borough'] == 'Scarborough'].reset_index(drop=True)
scarborough_data.head()
scarborough_data.shape

(17, 4)

We still have the same 4 columns, but now only 17 rows - the neighborhoods that are in Scarborough.

As we did with all of Toronto, let's visualize Scarborough and the neighborhoods within it.

In [ ]:
# create map of Scarborough using latitude and longitude values
map_scarborough = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(scarborough_data['Latitude'], scarborough_data['Longitude'], scarborough_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_scarborough)  
    
map_scarborough

#Ok.  17 neighborhoods, 17 blue dots.  We've made our map of Scarborough.

Now, we will utilize the Foursquare API to explore and segment the neighborhoods.

In [ ]:
CLIENT_ID = 'VDRGSGLETOQNKZOHJEUQSRGRSE1NJSGGL5THGUCSVL5TFXAR' # your Foursquare ID
CLIENT_SECRET = '02DGJ1EEZODRT3V3LARHRHUASN1ZPCSNXT4VNCH11KZ4VKHW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VDRGSGLETOQNKZOHJEUQSRGRSE1NJSGGL5THGUCSVL5TFXAR
CLIENT_SECRET:02DGJ1EEZODRT3V3LARHRHUASN1ZPCSNXT4VNCH11KZ4VKHW


###Let's explore a neighborhood in our dataframe.

Get the neighborhood's name.  I've chosen Woburn at index 3, just because the name caught my eye.

In [ ]:
scarborough_data.loc[3, 'Neighborhood']

'Woburn'

Get the neighborhood's latitude and longitude values.

In [ ]:
neighborhood_latitude = scarborough_data.loc[3, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = scarborough_data.loc[3, 'Longitude'] # neighborhood longitude value
neighborhood_name = scarborough_data.loc[3, 'Neighborhood'] # neighborhood name

print('Latitude and Longitude values of {} are {}, {}.'.format(neighborhood_name, neighborhood_latitude, neighborhood_longitude))


Latitude and Longitude values of Woburn are 43.7709921, -79.21691740000001.


Now, let's get the top 100 venues that are within a radius of 500 meters.

In [ ]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=VDRGSGLETOQNKZOHJEUQSRGRSE1NJSGGL5THGUCSVL5TFXAR&client_secret=02DGJ1EEZODRT3V3LARHRHUASN1ZPCSNXT4VNCH11KZ4VKHW&v=20180605&ll=43.7709921,-79.21691740000001&radius=500&limit=100'

Send the GET request and examine the results

In [ ]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6079bbb7c4b698483099c5d1'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4cc1d28c06c254815ac18547-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d1e0931735',
         'name': 'Coffee Shop',
         'pluralName': 'Coffee Shops',
         'primary': True,
         'shortName': 'Coffee Shop'}],
       'id': '4cc1d28c06c254815ac18547',
       'location': {'address': '300 Borough Dr',
        'cc': 'CA',
        'city': 'Scarborough',
        'country': 'Canada',
        'crossStreet': 'Scarborough Town Centre',
        'distance': 356,
        'formattedAddress': ['300 Borough Dr (Scarborough Town Centre)',
         'Scarborough ON',
         'Canada'],
   

All the information is in the *items* key.  So let's borrow the get_category_type function from a prior lab.

In [ ]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [ ]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name             categories        lat        lng
0           Starbucks            Coffee Shop  43.770037 -79.221156
1         Tim Hortons            Coffee Shop  43.770827 -79.223078
2  Korean Grill House  Korean BBQ Restaurant  43.770812 -79.214502

And how many venues were returned by Foursquare?

In [ ]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare.


##Let's explore neighborhoods in Scarborough

Let's create a function to repeat the same process for all 17 neighborhoods.

In [ ]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now let's write the code to run the above function on each neighborhood and create a new dataframe called scarborough_venues.


In [ ]:

scarborough_venues = getNearbyVenues(names=scarborough_data['Neighborhood'],
                                   latitudes=scarborough_data['Latitude'],
                                   longitudes=scarborough_data['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge


Let's check the size of the resulting dataframe

In [ ]:
print(scarborough_venues.shape)
scarborough_venues.head()

(84, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy’s       43.807448   
1              -79.160497  Royal Canadian Legion       43.782533   
2              -79.188711         RBC Royal Bank       43.766790   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711             Sail Sushi       43.765951   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.191151                  Bank  
3       -79.191537     Electronics Store  
4       -79.191275            Restaurant

Let's check how many venues were returned for each neighborhood.

In [ ]:
scarborough_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Birch Cliff, Cliffside West                                             4   
Cedarbrae                                                               8   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         3   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Golden Mile, Clairlea, Oakridge                                         8   
Guildwood, Morningside, West Hill                                       9   
Kennedy Park, Ionview, East Birchmount Park                             5   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor...                      3   
Rouge Hill, Port Union, Highland Creek                                  1   
Scarborough Village                                                     2   
Steeles West, L'Amoreaux West                                          12   
Wexford, Maryvale                                                       4   
Woburn                                                                  3   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Agincourt                                                                4   
Birch Cliff, Cliffside West                                              4   
Cedarbrae                                                                8   
Clarks Corners, Tam O'Shanter, Sullivan                                 12   
Cliffside, Cliffcrest, Scarborough Village West                          3   
Dorset Park, Wexford Heights, Scarborough Town ...                       5   
Golden Mile, Clairlea, Oakridge                                          8   
Guildwood, Morningside, West Hill                                        9   
Kennedy Park, Ionview, East Birchmount Park                              5   
Malvern, Rouge                                                           1   
Milliken, Agincourt North, Steeles East, L'Amor...                       3   
Rouge Hill, Port Union, Highland Creek                                   1   
Scarborough Village                                                      2   
Steeles West, L'Amoreaux West                                           12   
Wexford, Maryvale                                                        4   
Woburn                                                                   3   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Agincourt                                               4               4   
Birch Cliff, Cliffside West                             4               4   
Cedarbrae                                               8               8   
Clarks Corners, Tam O'Shanter, Sullivan                12              12   
Cliffside, Cliffcrest, Scarborough Village West         3               3   
Dorset Park, Wexford Heights, Scarborough Town ...      5               5   
Golden Mile, Clairlea, Oakridge                         8               8   
Guildwood, Morningside, West Hill                       9               9   
Kennedy Park, Ionview, East Birchmount Park             5               5   
Malvern, Rouge                                          1               1   
Milliken, Agincourt North, Steeles East, L'Amor...      3               3   
Rouge Hill, Port Union, Highland Creek                  1               1   
Scarborough Village                                     2               2   
Steeles West, L'Amoreaux West                          12              12   
Wexford, Maryvale           

Let's see how many unique categories can be curated from all the returned venues.

In [ ]:
print('There are {} unique categories.'.format(len(scarborough_venues['Venue Category'].unique())))

There are 51 unique categories.


##Analyzing each neighborhood

In [ ]:
# one hot encoding
scar_onehot = pd.get_dummies(scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
scar_onehot['Neighborhood'] = scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [scar_onehot.columns[-1]] + list(scar_onehot.columns[:-1])
scar_onehot = scar_onehot[fixed_columns]

scar_onehot.head()

Neighborhood  American Restaurant  \
0                          Malvern, Rouge                    0   
1  Rouge Hill, Port Union, Highland Creek                    0   
2       Guildwood, Morningside, West Hill                    0   
3       Guildwood, Morningside, West Hill                    0   
4       Guildwood, Morningside, West Hill                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    1               0   
2                   0            0       0     1    0               0   
3                   0            0       0     0    0               0   
4                   0            0       0     0    0               0   

   Bus Line  Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0         0            0     0                     0                   0   
1         0            0     0                     0                   0   
2         0            0     0                     0                   0   
3         0            0     0                     0                   0   
4         0            0     0                     0                   0   

   Clothing Store  Coffee Shop  College Stadium  Convenience Store  \
0               0            0                0                  0   
1               0            0                0                  0   
2               0            0                0                  0   
3               0            0                0                  0   
4               0            0                0                  0   

   Department Store  Discount Store  Donut Shop  Electronics Store  \
0                 0               0           0                  0   
1                 0               0           0                  0   
2                 0               0           0                  0   
3                 0               0           0                  1   
4                 0               0           0                  0   

   Fast Food Restaurant  Fried Chicken Joint  Gas Station  \
0                     1                    0            0   
1                     0                    0            0   
2                     0                    0            0   
3                     0                    0            0   
4                     0                    0            0   

   General Entertainment  Hakka Restaurant  Hobby Shop  Ice Cream Shop  \
0                      0                 0           0               0   
1                      0                 0           0               0   
2                      0                 0           0               0   
3                      0                 0           0               0   
4                      0                 0           0               0   

   Indian Restaurant  Intersection  Italian Restaurant  Jewelry Store  \
0                  0             0                   0              0   
1                  0             0                   0              0   
2                  0             0                   0              0   
3                  0             0                   0              0   
4                  0             0                   0              0   

   Korean BBQ Restaurant  Latin American Restaurant  Lounge  Medical Center  \
0                      0                          0       0               0   
1                      0                          0       0               0   
2                      0                          0       0               0   
3                      0                          0       0               0   
4                      0                          0       0               0   

   Mexican Restaurant  Middle Eastern Restaurant  Motel  Noodle House  Park  \
0                   0                          0      0             0     0   
1         

Let's examine the new dataframe size.

In [ ]:
scar_onehot.shape

(84, 52)

Next, let's group rows by neighborhood and by taking the mean of the frequency of ocurrence of each category.

In [ ]:
scar_grouped = scar_onehot.groupby('Neighborhood').mean().reset_index()
scar_grouped

Neighborhood  American Restaurant  \
0                                           Agincourt             0.000000   
1                         Birch Cliff, Cliffside West             0.000000   
2                                           Cedarbrae             0.000000   
3             Clarks Corners, Tam O'Shanter, Sullivan             0.000000   
4     Cliffside, Cliffcrest, Scarborough Village West             0.333333   
5   Dorset Park, Wexford Heights, Scarborough Town...             0.000000   
6                     Golden Mile, Clairlea, Oakridge             0.000000   
7                   Guildwood, Morningside, West Hill             0.000000   
8         Kennedy Park, Ionview, East Birchmount Park             0.000000   
9                                      Malvern, Rouge             0.000000   
10  Milliken, Agincourt North, Steeles East, L'Amo...             0.000000   
11             Rouge Hill, Port Union, Highland Creek             0.000000   
12                                Scarborough Village             0.000000   
13                      Steeles West, L'Amoreaux West             0.000000   
14                                  Wexford, Maryvale             0.000000   
15                                             Woburn             0.000000   

    Athletics & Sports  Auto Garage  Bakery      Bank  Bar  Breakfast Spot  \
0                0.000         0.00   0.000  0.000000  0.0        0.250000   
1                0.000         0.00   0.000  0.000000  0.0        0.000000   
2                0.125         0.00   0.125  0.125000  0.0        0.000000   
3                0.000         0.00   0.000  0.083333  0.0        0.000000   
4                0.000         0.00   0.000  0.000000  0.0        0.000000   
5                0.000         0.00   0.000  0.000000  0.0        0.000000   
6                0.000         0.00   0.250  0.000000  0.0        0.000000   
7                0.000         0.00   0.000  0.111111  0.0        0.111111   
8                0.000         0.00   0.000  0.000000  0.0        0.000000   
9                0.000         0.00   0.000  0.000000  0.0        0.000000   
10               0.000         0.00   0.000  0.000000  0.0        0.000000   
11               0.000         0.00   0.000  0.000000  1.0        0.000000   
12               0.000         0.00   0.000  0.000000  0.0        0.000000   
13               0.000         0.00   0.000  0.083333  0.0        0.083333   
14               0.000         0.25   0.250  0.000000  0.0        0.000000   
15               0.000         0.00   0.000  0.000000  0.0        0.000000   

    Bus Line  Bus Station  Café  Caribbean Restaurant  Chinese Restaurant  \
0       0.00        0.000  0.00                 0.000            0.000000   
1       0.00        0.000  0.25                 0.000            0.000000   
2       0.00        0.000  0.00                 0.125            0.000000   
3       0.00        0.000  0.00                 0.000            0.083333   
4       0.00        0.000  0.00                 0.000            0.000000   
5       0.00        0.000  0.00                 0.000            0.200000   
6       0.25        0.125  0.00                 0.000            0.000000   
7       0.00        0.000  0.00                 0.000            0.000000   
8       0.00        0.000  0.00                 0.000            0.200000   
9       0.00        0.000  0.00                 0.000            0.000000   
10      0.00        0.000  0.00                 0.000            0.000000   
11      0.00        0.000  0.00                 0.000            0.000000   
12      0.00        0.000  0.00                 0.000            0.000000   
13      0.00        0.000  0.00                 0.000            0.083333   
14      0.00        0.000  0.00                 0.000            0.000000   
15      0.00        0.000  0.00                 0.000            0.000000   

    Clothing Store  Coffee Shop  College Stadium  Convenience Store  \
0     

Let's confirm the new size

In [ ]:
scar_grouped.shape

(16, 52)

Let's print each neighborhood along with the top 5 most common venues.

In [ ]:
num_top_venues = 5

for hood in scar_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = scar_grouped[scar_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1             Breakfast Spot  0.25
2                     Lounge  0.25
3             Clothing Store  0.25
4        American Restaurant  0.00


----Birch Cliff, Cliffside West----
                   venue  freq
0           Skating Rink  0.25
1  General Entertainment  0.25
2                   Café  0.25
3        College Stadium  0.25
4    American Restaurant  0.00


----Cedarbrae----
                venue  freq
0     Thai Restaurant  0.12
1              Bakery  0.12
2                Bank  0.12
3  Athletics & Sports  0.12
4    Hakka Restaurant  0.12


----Clarks Corners, Tam O'Shanter, Sullivan----
                venue  freq
0         Pizza Place  0.17
1        Intersection  0.08
2        Noodle House  0.08
3   Convenience Store  0.08
4  Italian Restaurant  0.08


----Cliffside, Cliffcrest, Scarborough Village West----
                 venue  freq
0  American Restaurant  0.33
1         Intersection  0.

#Let's put that into a pandas dataframe

  First, let's write a function to sort the venues in descending order.

In [ ]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

  # create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = scar_grouped['Neighborhood']

for ind in np.arange(scar_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(scar_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  \
0                                        Agincourt         Clothing Store   
1                      Birch Cliff, Cliffside West  General Entertainment   
2                                        Cedarbrae        Thai Restaurant   
3          Clarks Corners, Tam O'Shanter, Sullivan            Pizza Place   
4  Cliffside, Cliffcrest, Scarborough Village West    American Restaurant   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant                Lounge   
1          Skating Rink                       Café       College Stadium   
2    Athletics & Sports                     Bakery                  Bank   
3    Chinese Restaurant               Noodle House  Fast Food Restaurant   
4          Intersection                      Motel           Auto Garage   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Coffee Shop   Fried Chicken Joint  Fast Food Restaurant   
1        Clothing Store           Gas Station   Fried Chicken Joint   
2           Gas Station   Fried Chicken Joint  Caribbean Restaurant   
3       Thai Restaurant          Intersection    Italian Restaurant   
4           Coffee Shop   Fried Chicken Joint  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0     Electronics Store            Donut Shop         Discount Store  
1  Fast Food Restaurant     Electronics Store             Donut Shop  
2      Hakka Restaurant     Convenience Store  General Entertainment  
3     Convenience Store           Gas Station    Fried Chicken Joint  
4    Athletics & Sports     Electronics Store             Donut Shop

#4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [ ]:
#set number of clusters
kclusters = 5

scar_grouped_clustering = scar_grouped.drop('Neighborhood', 1)

#run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(scar_grouped_clustering)

#check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 2], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [ ]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

scar_merged = scarborough_data

#merge scar_grouped with scar_data to add latitude/longitude for each neighborhood
scar_merged = scar_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

scar_merged.head()

Borough                            Neighborhood   Latitude  Longitude  \
0  Scarborough                          Malvern, Rouge  43.806686 -79.194353   
1  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711   
3  Scarborough                                  Woburn  43.770992 -79.216917   
4  Scarborough                               Cedarbrae  43.773136 -79.239476   

   Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0             2.0  Fast Food Restaurant  Vietnamese Restaurant   
1             0.0                   Bar  Vietnamese Restaurant   
2             1.0        Medical Center     Mexican Restaurant   
3             4.0           Coffee Shop  Korean BBQ Restaurant   
4             1.0       Thai Restaurant     Athletics & Sports   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Clothing Store           Gas Station   Fried Chicken Joint   
1         Clothing Store           Gas Station   Fried Chicken Joint   
2      Electronics Store            Donut Shop            Restaurant   
3  Vietnamese Restaurant      Hakka Restaurant           Gas Station   
4                 Bakery                  Bank           Gas Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Electronics Store            Donut Shop        Discount Store   
1  Fast Food Restaurant     Electronics Store            Donut Shop   
2   Rental Car Location                  Bank        Breakfast Spot   
3   Fried Chicken Joint  Fast Food Restaurant     Electronics Store   
4   Fried Chicken Joint  Caribbean Restaurant      Hakka Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store      Convenience Store  
1        Discount Store       Department Store  
2          Intersection      Convenience Store  
3            Donut Shop         Discount Store  
4     Convenience Store  General Entertainment

Lastly, let's visualize the resulting clusters.

In [ ]:
#create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

#set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0,1,len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(scar_merged['Latitude'], scar_merged['Longitude'], scar_merged['Neighborhood'], scar_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters



#5. Examine Clusters

Now, let's examine each cluster to determine the discriminating venue categories that distinguish each cluster.  Based on the defining categories, you can then assign a name to each cluster.

In [ ]:
scar_merged.loc[scar_merged['Cluster Labels'] == 0, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]




Neighborhood 1st Most Common Venue  \
1  Rouge Hill, Port Union, Highland Creek                   Bar   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1  Vietnamese Restaurant        Clothing Store           Gas Station   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1   Fried Chicken Joint  Fast Food Restaurant     Electronics Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1            Donut Shop        Discount Store       Department Store

In [ ]:
scar_merged.loc[scar_merged['Cluster Labels'] == 1, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]



Neighborhood  1st Most Common Venue  \
2                   Guildwood, Morningside, West Hill         Medical Center   
4                                           Cedarbrae        Thai Restaurant   
6         Kennedy Park, Ionview, East Birchmount Park             Hobby Shop   
7                     Golden Mile, Clairlea, Oakridge                 Bakery   
8     Cliffside, Cliffcrest, Scarborough Village West    American Restaurant   
9                         Birch Cliff, Cliffside West  General Entertainment   
10  Dorset Park, Wexford Heights, Scarborough Town...      Indian Restaurant   
11                                  Wexford, Maryvale            Auto Garage   
12                                          Agincourt         Clothing Store   
13            Clarks Corners, Tam O'Shanter, Sullivan            Pizza Place   
15                      Steeles West, L'Amoreaux West            Pizza Place   

    2nd Most Common Venue      3rd Most Common Venue  \
2      Mexican Restaurant          Electronics Store   
4      Athletics & Sports                     Bakery   
6             Coffee Shop             Discount Store   
7                Bus Line             Ice Cream Shop   
8            Intersection                      Motel   
9            Skating Rink                       Café   
10  Vietnamese Restaurant                  Pet Store   
11         Sandwich Place                     Bakery   
12         Breakfast Spot  Latin American Restaurant   
13     Chinese Restaurant               Noodle House   
15                   Bank               Intersection   

        4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
2                  Donut Shop             Restaurant   Rental Car Location   
4                        Bank            Gas Station   Fried Chicken Joint   
6            Department Store     Chinese Restaurant                  Bank   
7                Intersection            Bus Station                  Park   
8                 Auto Garage            Coffee Shop   Fried Chicken Joint   
9             College Stadium         Clothing Store           Gas Station   
10         Chinese Restaurant                 Bakery                  Bank   
11  Middle Eastern Restaurant  Vietnamese Restaurant        Clothing Store   
12                     Lounge            Coffee Shop   Fried Chicken Joint   
13       Fast Food Restaurant        Thai Restaurant          Intersection   
15             Breakfast Spot         Discount Store              Pharmacy   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
2                    Bank        Breakfast Spot          Intersection   
4    Caribbean Restaurant      Hakka Restaurant     Convenience Store   
6                     Bar           Gas Station   Fried Chicken Joint   
7   Vietnamese Restaurant       College Stadium   Fried Chicken Joint   
8    Fast Food Restaurant    Athletics & Sports     Electronics Store   
9     Fried Chicken Joint  Fast Food Restaurant     Electronics Store   
10    Fried Chicken Joint    Athletics & Sports  Fast Food Restaurant   
11    Fried Chicken Joint  Fast Food Restaurant     Electronics Store   
12   Fast Food Restaurant     Electronics Store            Donut Shop   
13     Italian Restaurant     Convenience Store           Gas Station   
15            Coffee Shop    Chinese Restaurant     Electronics Store   

   10th Most Common Venue  
2       Convenience Store  
4   General Entertainment  
6    Fast Food Restaurant  
7    Fast Food Restaurant  
8              Donut Shop  
9              Donut Shop  
10      Electronics Store  
11             Donut Shop  
12         Discount Store  
13    Fried Chicken Joint  
15         Sandwich Place

In [ ]:
scar_merged.loc[scar_merged['Cluster Labels'] == 2, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]


Neighborhood 1st Most Common Venue  2nd Most Common Venue  \
0  Malvern, Rouge  Fast Food Restaurant  Vietnamese Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Clothing Store           Gas Station   Fried Chicken Joint   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Electronics Store            Donut Shop        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
0      Department Store      Convenience Store

In [ ]:
scar_merged.loc[scar_merged['Cluster Labels'] == 4, scar_merged.columns[[1] + list(range(5, scar_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  2nd Most Common Venue  \
3       Woburn           Coffee Shop  Korean BBQ Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3  Vietnamese Restaurant      Hakka Restaurant           Gas Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3   Fried Chicken Joint  Fast Food Restaurant     Electronics Store   

  9th Most Common Venue 10th Most Common Venue  
3            Donut Shop         Discount Store